contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+    coverNFT = _coverNFT;
+
+    stakingPoolProxyCodeHash = CoverUtilsLib.calculateProxyCodeHash();
+    stakingPoolImplementation = _stakingPoolImplementation;
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+
+  /// @dev Migrates covers from V1. Meant to be used by Claims.sol and Gateway.sol to allow the
+  /// users of distributor contracts to migrate their NFTs.
+  ///
+  /// @param coverId     V1 cover identifier
+  /// @param fromOwner   The address from where this function is called that needs to match the
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function migrateCoverFromOwner(
+    uint coverId,
+    address fromOwner,
+    address toNewOwner
+  ) external override onlyInternal {
+    _migrateCoverFromOwner(coverId, fromOwner, toNewOwner);
+  }
+
+  /// @dev Migrates covers from V1
+  ///
+  /// @param coverId     V1 cover identifier
+  /// @param fromOwner   The address from where this function is called that needs to match the
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function _migrateCoverFromOwner(
+    uint coverId,
+    address fromOwner,
+    address toNewOwner
+  ) internal {
+
+    CoverUtilsLib.migrateCoverFromOwner(
+      CoverUtilsLib.MigrateParams(
+        coverId,
+        fromOwner,
+        toNewOwner,
+        ICoverNFT(coverNFT),
+        quotationData,
+        tokenController(),
+        productsV1
+      ),
+      _products,
+      _productTypes,
+      _coverData,
+      _coverSegments
+    );
+  }
+
+  /// @dev Migrates covers from V1. Meant to be used by EOA Nexus Mutual members
+  ///
+  /// @param coverIds    Legacy (V1) cover identifiers
+  /// @param toNewOwner  The address for which the V2 cover NFT is minted
+  function migrateCovers(uint[] calldata coverIds, address toNewOwner) external override {
+    for (uint i = 0; i < coverIds.length; i++) {
+      _migrateCoverFromOwner(coverIds[i], msg.sender, toNewOwner);
+    }
+  }
+
+  function buyCover(
+    BuyCoverParams memory params,
+    PoolAllocationRequest[] memory allocationRequests
+  ) external payable override onlyMember whenNotPaused returns (uint /*coverId*/) {
+
+    require(_products.length > params.productId, "Cover: Product not found");
+    Product memory product = _products[params.productId];
+    require(product.initialPriceRatio != 0, "Cover: Product not initialized");
+    require(
+      isAssetSupported(product.coverAssets, params.payoutAsset),
+      "Cover: Payout asset is not supported"
+    );
+    require(params.period >= MIN_COVER_PERIOD, "Cover: Cover period is too short");
+    require(params.period <= MAX_COVER_PERIOD, "Cover: Cover period is too long");
+    require(params.commissionRatio <= MAX_COMMISSION_RATIO, "Cover: Commission rate is too high");
+
+    uint totalPremiumInNXM = _buyCover(params, _coverData.length, allocationRequests);
+
+    IPool _pool = pool();
+    uint tokenPriceInPaymentAsset = _pool.getTokenPrice(params.paymentAsset);
```

Comment:
Exactly the same call is perform inside `_buyCover`. Pass `tokenPriceInPaymentAsset` as a parameter to `_buyCover` instead